src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1420 @@
+            int[][] w0 = new int[mlDsa_k][mlDsa_n];
+            int[][] w1 = new int[mlDsa_k][mlDsa_n];
+            decompose(w, w0, w1);
+            mlDsaVectorInverseNtt(y);
+
+            //Get commitment hash
+            hash.update(mu);
+            hash.update(simpleBitPack(wCoeffSize, w1));
+            commitmentHash = hash.squeeze(lambda/4);
+            hash.reset();
+
+            //Get z and r0
+            int[] c = sampleInBall(commitmentHash);
+            mlDsaNtt(c); //c is now in NTT domain
+            int[][] cs1 = nttConstMultiply(c, s1);
+            int[][] cs2 = nttConstMultiply(c, s2);
+            mlDsaVectorInverseNtt(cs1);
+            mlDsaVectorInverseNtt(cs2);
+            z = vectorAdd(y, cs1);
+            int[][] r0 = vectorSub(w0, cs2);
+
+            //Update z and h
+            kappa += mlDsa_l;
+            if (vectorNormBound(z, gamma1 - beta) || vectorNormBound(r0, gamma2 - beta)) {
+                continue;
+            } else {
+                int[][] ct0 = nttConstMultiply(c, t0);
+                mlDsaVectorInverseNtt(ct0);
+                h = makeHint(vectorConstMul(-1, ct0), vectorAdd(vectorAdd(w, vectorConstMul(-1, cs2)), ct0));
+                if (vectorNormBound(ct0, gamma2) || (hammingWeight(h) > omega)) {
+                    continue;
+                }
+            }
+            return new ML_DSA_Signature(commitmentHash, z, h);
+        }
+    }
+
+    public boolean verify(ML_DSA_PublicKey pk, byte[] message, ML_DSA_Signature sig) {
+        var hash = new SHAKE256(0);
+
+        //Check hint encoding
+        if (sig.hint() == null) {return false;}
+
+        //Expand A
+        int[][][] aHat = generateA(pk.rho());
+
+        //Generate tr
+        hash.update(pkEncode(pk));
+        byte[] tr = hash.squeeze(mlDsaTrLength);
+        hash.reset();
+
+        //Generate mu
+        hash.update(tr);
+        hash.update(message);
+        byte[] mu = hash.squeeze(mlDsaMuLength);
+        hash.reset();
+
+        //Get verifiers challenge
+        int[] cHat = sampleInBall(sig.commitmentHash());
+        mlDsaNtt(cHat);
+
+        //Deal with z
+        int[][] z = Arrays.stream(sig.response()).map(int[]::clone).toArray(int[][]::new);
+        boolean zNorm = vectorNormBound(z, gamma1 - beta);
+        mlDsaVectorNtt(z);
+
+        //Reconstruct signer's commitment
+        int[][] aHatZ = matrixVectorPointwiseMultiply(aHat, z);
+        int[][] t1Hat = vectorConstMul(1 << mlDsa_d, pk.t1());
+        mlDsaVectorNtt(t1Hat);
+        int[][] wApprox = vectorSub(aHatZ,nttConstMultiply(cHat, t1Hat));
+        mlDsaVectorInverseNtt(wApprox);
+        int[][] w1Prime = useHint(sig.hint(), wApprox);
+
+        //Hash signer's commitment
+        hash.update(mu);
+        hash.update(simpleBitPack(wCoeffSize, w1Prime));
+        byte[] cTildePrime = hash.squeeze(lambda/4);
+
+        //Check verify conditions
+        boolean hashEq = Arrays.equals(sig.commitmentHash(), cTildePrime);
```

Comment:
Do we want a constant-time array check here and use `MessageDigest.isEqual()` instead?